firewood/src/merkle/parallel.rs

```diff
@@ -0,0 +1,500 @@
+                    return Ok(Some(new_root));
+                }
+
+                // Return the child as the new root. Update its partial path to include the index value.
+                let mut child = match child {
+                    Child::Node(child_node) => std::mem::take(child_node),
+                    Child::AddressWithHash(addr, _) => nodestore.read_for_update((*addr).into())?,
+                    Child::MaybePersisted(maybe_persisted, _) => {
+                        nodestore.read_for_update(maybe_persisted.clone())?
+                    }
+                };
+
+                // The child's partial path is the concatenation of its (now removed) parent, which
+                // should always be empty because of our prepare step, its (former) child index, and
+                // its partial path. Because the parent's partial path should always be empty, we
+                // replace it here with a Path::new().
+                let partial_path = Path::from_nibbles_iterator(
+                    Path::new()
+                        .iter()
+                        .copied()
+                        .chain(once(child_index as u8))
+                        .chain(child.partial_path().iter().copied()),
+                );
+
+                match child {
+                    Node::Branch(ref mut child_branch) => {
+                        child_branch.partial_path = partial_path;
+                    }
+                    Node::Leaf(ref mut leaf) => {
+                        leaf.partial_path = partial_path;
+                    }
+                }
+                Ok(Some(child))
+            }
+        }
+    }
+
+    /// Creates a worker for performing operations on a subtrie, with the subtrie being determined
+    /// by the value of the `first_nibble`.
+    fn create_worker(
+        pool: &ThreadPool,
+        proposal: &NodeStore<MutableProposal, FileBacked>,
+        root_branch: &mut Box<BranchNode>,
+        first_nibble: u8,
+        worker_sender: Sender<Response<FileBacked>>,
+    ) -> Result<WorkerState, FileIoError> {
+        // Create a channel for the coordinator (main thread) to send messages to this worker.
+        let child_channel = mpsc::channel();
+
+        // Get the child node from the proposal
+        let child = root_branch
+            .children
+            .get_mut(first_nibble as usize)
+            .expect("index error")
+            .take();
+
+        // Build a nodestore from the child node
+        let worker_nodestore = NodeStore::from_child(proposal, child)?;
+
+        // Spawn a worker from the threadpool for this nibble. The worker will send messages to the coordinator
+        // using `worker_sender`.
+        pool.spawn(move || {
+            let mut merkle = Merkle::from(worker_nodestore);
+            loop {
+                // Wait for a message on the receiver child channel. Break out of loop if there is an error.
+                let request = match child_channel.1.recv() {
```

Comment:
This made me scratch my head as to why this worked. The compiler was smart enough to realize that the only thing being moved here is `child_channel.1` leaving `child_channel` is a "partially moved" state. See the comment above on how to fix this.